src/transformers/generation_utils.py

```diff
@@ -638,9 +641,18 @@ def _get_logits_processor(
             processors.append(InfNanRemoveLogitsProcessor())
         return processors
 
-    def _get_stopping_criteria(self, max_length: Optional[int], max_time: Optional[float]) -> StoppingCriteriaList:
-        stopping_criteria = StoppingCriteriaList()
-        if max_length is not None:
+    def _get_stopping_criteria(
+        self,
+        max_length: Optional[int],
+        max_time: Optional[float],
+        max_new_tokens: Optional[int],
+        start_length: int,
+        stopping_criteria: Optional[StoppingCriteriaList],
+    ) -> StoppingCriteriaList:
+        if stopping_criteria is None:
+            stopping_criteria = StoppingCriteriaList()
+        max_length_in_criteria = any([isinstance(criteria, MaxLengthCriteria) for criteria in stopping_criteria])
+        if max_length is not None and not max_length_in_criteria:
```

Comment:
I'm fine with this I think -> it means that `stopping_criteria` is always more important than `max_length` which sounds good to me

Reply:
Actually thinking more about I think it would be cleaner to just have the following logic:

if `stopping_criteria` is provided than only this list is used and nothing else. In a first step this allows us some nasty nested use cases. IMO we could just check if `len(stopping_criteria) > 0` and if this is the case we don't even call the function `_get_stopping_criteria`. IMO someone that uses that functionality understands `generate()` quite well and doesn't need much magic under-the-hood.

Reply:
What do you think @Narsil ?

Reply:
That's an option that is definitely viable.

The core things I think are important:

- If a user specified something, we need to respect it
- If something comes as a default it cannot override anything user specified.
- If user specification are unclear/unsound, yell loud and clear about what is going on, and what the code is going to do to save the `generate`.

Option 1 (current):
```python
model.generate(...., no_repeat_n_tokens=3)
```
I need to add some even more clever functionality
```python
mylogitsProcessor=LogitsProcessorList(MyLogits())
model.generate(...., logits_processor=mylogitsProcessor, no_repeat_n_tokens=3)
```
(You can keep the easiness of `generate`).

Option 2 (logits_processor is a full override):
```python
model.generate(..., no_repeat_n_tokens=3)
```
becomes
```python
my_logits_processor = LogitsProcessorList(
   NoRepeatLogitsProcessor(3),
  MyLogits()
)
model.generate(..., logits_processor=logits_processor)
```

Option 1 has the advantage that we can keep some options simpler to use and still add some custom logits processor, if we're careful enough that no non-user defined variable can ever override the `logits_processor` in a hidden way, then we're good to go. (Only user defined arguments are able to modify). This is ofc a guarantee that might be tricky to keep in the future, so we are taking a risk of silently breaking things. It also makes user code trickier to understand since there is no ONE way to define logits_processor, so it might lead to hard to understand behavior.


Option 2 has the advantage that it has a single point of definition of logits processor. The disadvantage is that it requires more changes on the user the first time he uses this variable. We also need to yell quite strongly that we're simply ignoring every other argument, which might not be obvious (we could even crash at this point, since it's a really easy to overlook thing and will definitely yield poor results). 

IMHO, both are really fine, we just need to stick to one option. I suggested to @lvwerra Option 1, because I thought that `max_length` was the only variable that was always defined even if not user supplied. If this assumption is wrong and hard to make on a small list of variable, then I don't think we should stick with Option 1. Option 2 does have drawbacks as we're rejecting complexity back into user code instead of absorbing it like we're doing right now. But it does help separation of concern.

Reply:
Thanks for the great write-up! Actually there are a bunch of things that are always defined (top_k is always defined e.g.) but then also lots of models always define `num_beams` in their config alongside other parameters so I would very much prefer option 2 here.

Reply:
@lvwerra are you ok switching to Option 2, seems my assumption was incorrect :( ?